File: 2022/P02/problem2b.py
data = []
with open("problem2.txt", 'r') as f:
    data = f.read().strip('\n').split('\n')

def f(char):
    if char == 'A' or char == 'X':
        return 1
    if char == 'B' or char == 'Y':
        return 2
    if char == 'C' or char == 'Z':
        return 3

score = 0
for p in data:
    chars = p.split(' ')
    nums = [f(x) for x in chars]
    round_score = 0
    if nums[1] == 1:
        round_score = (nums[0] - 1)
        if round_score == 0:
            round_score = 3
    elif nums[1] == 2:
        round_score = 3 + nums[0]
    elif nums[1] == 3:
        round_score = 6 + (nums[0] % 3) + 1
    score += round_score

print(score)

File: 2023/P02/problem2b.rs
use std::fs;
use std::cmp;

fn main() {
    let data = fs::read_to_string("problem2.txt")
        .expect("Failed to read file");
    let data = data.trim()
        .split('\n');
    let mut sum = 0;
    for el in data {
        let games = el.trim().split(": ").collect::<Vec<_>>()[1];
        let mut r_max = 0;
        let mut g_max = 0;
        let mut b_max = 0;
        for g in games.split("; ") {
            let vals = g.split(", ");
            for v_col in vals {
                let vec = v_col.split(" ").collect::<Vec<_>>();
                let v = vec[0];
                let col = vec[1]; 
                match v.parse::<u32>() {
                    Ok(num) => {
                        match col {
                            "red" => {
                                r_max = cmp::max(r_max, num);
                            },
                            "green" => {
                                g_max = cmp::max(g_max, num);
                            },
                            "blue" => {
                                b_max = cmp::max(b_max, num);
                            },
                            _ => {}
                        }
                    },
                    _ => {}
                }
            }
        }
        sum += r_max * g_max * b_max;
    }
    println!("{sum}");
}
